rusk/src/lib/lib.rs

```diff
@@ -34,97 +47,427 @@ pub static PUB_PARAMS: Lazy<PublicParameters> = Lazy::new(|| unsafe {
 
 const STREAM_BUF_SIZE: usize = 64;
 
-pub struct RuskBuilder {
-    id: Option<NetworkStateId>,
-    path: Option<PathBuf>,
-    backend: fn() -> BackendCtor<DiskBackend>,
+pub struct RuskInner {
+    pub current_commit: CommitId,
+    pub base_commit: CommitId,
+    pub vm: VM,
 }
 
-impl RuskBuilder {
-    pub fn new(backend: fn() -> BackendCtor<DiskBackend>) -> Self {
-        Self {
-            id: None,
-            path: None,
-            backend,
+#[derive(Clone)]
+pub struct Rusk {
+    inner: Arc<Mutex<RuskInner>>,
+    dir: PathBuf,
+    stream_buffer_size: usize,
+}
+
+impl Rusk {
+    pub fn new<P: AsRef<Path>>(dir: P) -> Result<Self> {
+        let dir = dir.as_ref();
+        let commit_id_path = to_rusk_state_id_path(dir);
+
+        let base_commit = CommitId::restore(commit_id_path)?;
+        let mut vm = VM::new(dir)?;
+
+        let mut session = vm.session();
+        session.restore(&base_commit)?;
+
+        rusk_abi::register_host_queries(&mut vm);
+
+        let inner = Arc::new(Mutex::new(RuskInner {
+            current_commit: base_commit,
+            base_commit,
+            vm,
+        }));
+
+        Ok(Self {
+            inner,
+            dir: dir.into(),
+            stream_buffer_size: STREAM_BUF_SIZE,
+        })
+    }
+
+    pub fn execute_transactions(
+        &self,
+        block_height: u64,
+        block_gas_limit: u64,
+        generator: BlsPublicKey,
+        txs: Vec<Transaction>,
+    ) -> Result<(Vec<SpentTransaction>, Vec<Transaction>, [u8; 32])> {
+        let mut inner = self.inner.lock();
+
+        let mut session = inner.vm.session();
+        rusk_abi::set_block_height(&mut session, block_height);
+
+        let mut block_gas_left = block_gas_limit;
+
+        let mut spent_txs = Vec::with_capacity(txs.len());
+        let mut discarded_txs = Vec::with_capacity(txs.len());
+
+        let mut dusk_spent = 0;
+
+        let mut nullifiers = BTreeSet::new();
+
+        // Here we discard transactions that:
+        // - Fail parsing
+        // - Use nullifiers that are already use by previous TXs
+        // - Spend more gas than the running `block_gas_left`
```

Comment:
No.
Discarded transaction are transaction that are not processable (indeed, the node will remove them from the mempool)
Said that, transactions which spend more gas than the running `block_gas_left` should not added to `discarded_tx` neither to `spent_transaction`.

Reply:
Fixed in #856